File: system/core/auth/permissions.py
from rest_framework import permissions
from django.core.exceptions import PermissionDenied

class IsAdultoMayor(permissions.BasePermission):
    """
    Permiso para usuarios con rol Adulto Mayor.
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.has_role('ADULTO_MAYOR')

class IsEstudianteUCT(permissions.BasePermission):
    """
    Permiso para usuarios con rol Estudiante UCT.
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.has_role('ESTUDIANTE_UCT')

class IsAdministrador(permissions.BasePermission):
    """
    Permiso para usuarios con rol Administrador.
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.has_role('ADMINISTRADOR')

class CanPublishSolicitud(permissions.BasePermission):
    """
    Permiso para publicar solicitudes (solo Adulto Mayor).
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.has_role('ADULTO_MAYOR')

class CanAcceptTask(permissions.BasePermission):
    """
    Permiso para aceptar tareas (solo Estudiante UCT).
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.has_role('ESTUDIANTE_UCT')

class CanCloseTask(permissions.BasePermission):
    """
    Permiso para cerrar tareas (Adulto Mayor o Estudiante UCT asignado).
    """
    def has_object_permission(self, request, view, obj):
        if not request.user.is_authenticated:
            return False
        # Adulto Mayor puede cerrar si es el presidente
        if request.user.has_role('ADULTO_MAYOR') and obj.presidente == request.user:
            return True
        # Estudiante UCT puede cerrar si está asignado
        if request.user.has_role('ESTUDIANTE_UCT') and request.user in obj.voluntarios.all():
            return True
        return False

class CanModerateContent(permissions.BasePermission):
    """
    Permiso para moderar contenido (solo Administrador).
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.has_role('ADMINISTRADOR')

class CanGenerateCertificate(permissions.BasePermission):
    """
    Permiso para generar certificados (Estudiante UCT emite, Administrador valida).
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated and (
            request.user.has_role('ESTUDIANTE_UCT') or request.user.has_role('ADMINISTRADOR')
        )

class CanViewSensitiveData(permissions.BasePermission):
    """
    Permiso para ver datos sensibles como dirección (solo cuando asignado a tarea).
    """
    def has_object_permission(self, request, view, obj):
        if not request.user.is_authenticated:
            return False
        # Solo mostrar dirección si la tarea está asignada y el usuario es el voluntario asignado
        if obj.estado == 'ASIGNADA' and request.user in obj.voluntarios.all():
            return True
        return False

File: server/src/presentation/middleware/permissions.js
// ============================================================================
// Middleware: Role-Based Permissions — RNF-SEG
// Equivalente a core/auth/permissions.py de Django (DRF BasePermission)
// Permisos granulares por rol para cada operación
// ============================================================================

/**
 * Middleware factory: Restringe acceso por rol(es).
 * Equivalente a IsAdultoMayor, IsEstudianteUCT, IsAdministrador de Django.
 * @param  {...string} roles - Roles permitidos
 */
function requireRole(...roles) {
  return (req, res, next) => {
    if (!req.user) {
      return res.status(401).json({ success: false, error: 'No autenticado.' });
    }
    if (!roles.includes(req.user.rol)) {
      return res.status(403).json({
        success: false,
        error: `Permiso denegado. Se requiere rol: ${roles.join(' o ')}.`,
      });
    }
    next();
  };
}

// ============================================================================
// Permisos específicos (equivalentes directos de Django permissions.py)
// ============================================================================

/**
 * CanPublishSolicitud — Solo Adulto Mayor o Tutor puede publicar solicitudes.
 * Equivalente a: class CanPublishSolicitud(permissions.BasePermission)
 */
const canPublishSolicitud = requireRole('ADULTO_MAYOR', 'TUTOR');

/**
 * CanAcceptTask — Solo Estudiante UCT puede aceptar tareas.
 * Equivalente a: class CanAcceptTask(permissions.BasePermission)
 */
const canAcceptTask = requireRole('ESTUDIANTE');

/**
 * CanModerateContent — Solo Administrador puede moderar contenido.
 * Equivalente a: class CanModerateContent(permissions.BasePermission)
 */
const canModerateContent = requireRole('ADMIN');

/**
 * CanGenerateCertificate — Estudiante o Admin pueden generar certificados.
 * Equivalente a: class CanGenerateCertificate(permissions.BasePermission)
 */
const canGenerateCertificate = requireRole('ESTUDIANTE', 'ADMIN');

/**
 * IsAdultoMayor — Permiso para usuarios con rol Adulto Mayor.
 */
const isAdultoMayor = requireRole('ADULTO_MAYOR');

/**
 * IsEstudianteUCT — Permiso para usuarios con rol Estudiante UCT.
 */
const isEstudianteUCT = requireRole('ESTUDIANTE');

/**
 * IsAdministrador — Permiso para usuarios con rol Administrador.
 */
const isAdministrador = requireRole('ADMIN');

module.exports = {
  requireRole,
  canPublishSolicitud,
  canAcceptTask,
  canModerateContent,
  canGenerateCertificate,
  isAdultoMayor,
  isEstudianteUCT,
  isAdministrador,
};
